bin/runtime/src/migrations/custom_contracts_migration.rs

```diff
@@ -0,0 +1,306 @@
+        storage_item_deposit: BalanceOf<T>,
+        storage_base_deposit: BalanceOf<T>,
+    }
+
+    const OLD_INFO_ENCODING_LEN: usize = 33 + 32 + 16;
+    const NEW_INFO_ENCODING_LEN: usize = 33 + 32 + 4 + 4 + 16 + 16 + 16;
+    const MAX_NUM_ITEMS_PROCESSED_PER_MIGRATION: u32 = 140000;
+
+    enum EitherContractInto<T: Config> {
+        Old(OldContractInfo<T>),
+        New(ContractInfo<T>),
+    }
+
+    #[storage_alias]
+    type ContractInfoOf<T: Config, V> =
+        StorageMap<Pallet<T>, Twox64Concat, <T as frame_system::Config>::AccountId, V>;
+
+    fn decode_to_either_contract_info<T: Config>(
+        input: &mut &[u8],
+    ) -> Result<EitherContractInto<T>, Error> {
+        let len = input.len();
+        match len {
+            OLD_INFO_ENCODING_LEN => {
+                if let Ok(c_info) = OldContractInfo::<T>::decode_all(input) {
+                    Ok(EitherContractInto::Old(c_info))
+                } else {
+                    Err(Error::from("Failed to decode ContractInfo to Old type"))
+                }
+            }
+            NEW_INFO_ENCODING_LEN => {
+                if let Ok(c_info) = ContractInfo::<T>::decode_all(input) {
+                    Ok(EitherContractInto::New(c_info))
+                } else {
+                    Err(Error::from("Failed to decode ContractInfo to New type"))
+                }
+            }
+            _ => Err(Error::from("Wrong size of ContractInfo")),
+        }
+    }
+
+    // Returns the used weight and whether this migration is the final one.
+    pub fn migrate<T: Config>() -> (Weight, bool) {
+        log::info!(
+            target: TARGET,
+            "Running v7->v8 migration of pallet contracts"
+        );
+        let mut weight = Weight::zero();
+        let mut total_bytes: u32 = 0;
+        let mut total_items: u32 = 0;
+        let mut num_old_contracts_migrated: u32 = 0;
+        let mut num_old_contracts_skipped: u32 = 0;
+        let mut num_new_contracts: u32 = 0;
+        let mut num_contracts_erred: u32 = 0;
+
+        // Note that we use V = ContractInfo<T> and not V = OldContractInfo<T>, but that does not matter for the prefix.
+        let prefix = ContractInfoOf::<T, ContractInfo<T>>::final_prefix();
+        let mut previous_key = prefix.clone().to_vec();
+        // Iterate over all elements of the ContractInfoOf map
+        while let Some(next) =
+            sp_io::storage::next_key(&previous_key).filter(|n| n.starts_with(&prefix))
+        {
+            previous_key = next;
+            let value = sp_io::storage::get(&previous_key)
+                .expect("The loop conditions checks for existence of the key; qed");
+            match decode_to_either_contract_info::<T>(&mut &value[..]) {
+                Ok(EitherContractInto::Old(old)) => {
+                    if total_items >= MAX_NUM_ITEMS_PROCESSED_PER_MIGRATION {
+                        num_old_contracts_skipped.saturating_accrue(1);
+                        continue;
```

Comment:
we could just break from the while loop here and loose information about num_old_contracts_skipped and num_new_contracts but gain some time on skipping the decoding of the rest contracts.

Reply:
Agreed, this could save a bit, but not a lot actually. In the best case scenario this saves around 4k reads, which is 3% of the typical number of 150k reads we make per migration. 

I like to see the full data logged in the summary, so that we can make sure everything is progressing as expected...